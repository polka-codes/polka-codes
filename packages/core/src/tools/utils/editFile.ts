// Core logic for flexible file editing with search/replace operations
// generated by polka.codes

/**
 * Edit operation using search/replace pattern
 */
export interface EditOperation {
  search: string
  replace: string
}

export const START_OF_FILE = '<<<START_OF_FILE>>>'
export const END_OF_FILE = '<<<END_OF_FILE>>>'

/**
 * Apply multiple edit operations to file content using search/replace pattern.
 * @param fileContent - The original file content
 * @param operations - Array of edit operations to apply
 * @returns The modified file content
 */
export const editFile = async (fileContent: string, operations: EditOperation[]): Promise<string> => {
  if (!operations || operations.length === 0) {
    throw new Error('At least one edit operation is required')
  }

  let updatedContent = fileContent

  // Process operations in order
  for (const operation of operations) {
    updatedContent = await applyEditOperation(updatedContent, operation)
  }

  return updatedContent
}

async function applyEditOperation(fileContent: string, operation: EditOperation): Promise<string> {
  const { search, replace } = operation

  // Validate operation
  if (search === START_OF_FILE && replace === END_OF_FILE) {
    throw new Error('Cannot search for START_OF_FILE and replace with END_OF_FILE')
  }

  if (search === END_OF_FILE && replace === START_OF_FILE) {
    throw new Error('Cannot search for END_OF_FILE and replace with START_OF_FILE')
  }

  // Handle special cases for start/end of file
  if (search === START_OF_FILE) {
    // Insert at start of file
    return replace + fileContent
  }

  if (search === END_OF_FILE) {
    // Insert at end of file
    return fileContent + replace
  }

  // Handle normal search and replace
  const index = fileContent.indexOf(search)
  if (index === -1) {
    throw new Error(`Could not find text: ${search}`)
  }

  // Replace first occurrence only
  return fileContent.slice(0, index) + replace + fileContent.slice(index + search.length)
}
