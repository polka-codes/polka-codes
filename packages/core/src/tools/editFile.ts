// Tool for flexible file editing with search/replace operations
// generated by polka.codes

import { type FullToolInfo, PermissionLevel, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'
import { getArray, getString } from './utils'
import { END_OF_FILE, type EditOperation, START_OF_FILE, editFile } from './utils/editFile'

export const toolInfo = {
  name: 'edit_file',
  description: 'Request to edit file contents using search/replace operations. Supports multiple edit operations in a single call.',
  parameters: [
    {
      name: 'path',
      description: 'The path of the file to edit',
      required: true,
      usageValue: 'File path here',
    },
    {
      name: 'operations',
      description: 'Edit operation with search and replace parameters',
      required: true,
      allowMultiple: true,
      children: [
        {
          name: 'search',
          description: `Text to search for and replace (use ${START_OF_FILE} to insert at file start, ${END_OF_FILE} to insert at file end)`,
          required: true,
          usageValue: 'Text to find and replace',
        },
        {
          name: 'replace',
          description: 'Text to replace the search text with',
          required: true,
          usageValue: 'Replacement text',
        },
      ],
      usageValue: 'operations here',
    },
  ],
  examples: [
    {
      description: 'Replace specific text',
      parameters: [
        {
          name: 'path',
          value: 'src/main.ts',
        },
        {
          name: 'operations',
          value: {
            search: 'function oldFunction() {\n  return 42;\n}',
            replace: 'function newFunction() {\n  return "new implementation";\n}',
          },
        },
      ],
    },
    {
      description: 'Insert at start of file',
      parameters: [
        {
          name: 'path',
          value: 'src/header.ts',
        },
        {
          name: 'operations',
          value: {
            search: START_OF_FILE,
            replace: '// File header comment\n',
          },
        },
      ],
    },
    {
      description: 'Multiple operations in one call',
      parameters: [
        {
          name: 'path',
          value: 'src/utils.ts',
        },
        {
          name: 'operations',
          value: [
            {
              search: 'import React from "react"',
              replace: 'import React, { useState } from "react"',
            },
            {
              search: 'function Component() {\n  return (',
              replace: 'function Component() {\n  const [state, setState] = useState(false);\n  return (',
            },
          ],
        },
      ],
    },
    {
      description: 'Append content at end of file',
      parameters: [
        {
          name: 'path',
          value: 'src/footer.ts',
        },
        {
          name: 'operations',
          value: {
            search: END_OF_FILE,
            replace: '\n// End of file appended comment\n',
          },
        },
      ],
    },
  ],
  permissionLevel: PermissionLevel.Write,
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.readFile || !provider.writeFile) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to edit file. Abort.',
    }
  }

  const path = getString(args, 'path')
  const operations = getArray(args, 'operations') as unknown as EditOperation[]

  if (!operations || operations.length === 0) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>At least one edit operation is required</error_message></error>`,
    }
  }

  const fileContent = await provider.readFile(path)

  if (fileContent == null) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>File not found</error_message></error>`,
    }
  }

  try {
    const result = await editFile(fileContent, operations)
    await provider.writeFile(path, result)

    return {
      type: ToolResponseType.Reply,
      message: `<edit_file_path>${path}</edit_file_path>`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>${error instanceof Error ? error.message : String(error)}</error_message></error>`,
    }
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.readFile && !!provider.writeFile
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfo
