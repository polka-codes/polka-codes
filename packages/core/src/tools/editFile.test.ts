// Tests for editFile tool
// generated by polka.codes

import { describe, expect, it } from 'bun:test'
import { handler, isAvailable, toolInfo } from './editFile'
import type { FilesystemProvider } from './provider'

describe('editFile tool', () => {
  describe('toolInfo', () => {
    it('should have correct tool definition', () => {
      expect(toolInfo).toMatchSnapshot()
    })
  })

  describe('isAvailable', () => {
    it('should return true when provider has readFile and writeFile', () => {
      const provider: FilesystemProvider = {
        readFile: async () => 'content',
        writeFile: async () => {},
      }
      expect(isAvailable(provider)).toBe(true)
    })

    it('should return false when provider lacks readFile', () => {
      const provider: FilesystemProvider = {
        writeFile: async () => {},
      }
      expect(isAvailable(provider)).toBe(false)
    })

    it('should return false when provider lacks writeFile', () => {
      const provider: FilesystemProvider = {
        readFile: async () => 'content',
      }
      expect(isAvailable(provider)).toBe(false)
    })
  })

  describe('handler', () => {
    const createProvider = (fileContent: string | undefined = 'test content'): FilesystemProvider => ({
      readFile: async () => fileContent,
      writeFile: async () => {},
    })

    it('should return error when provider lacks capabilities', async () => {
      const provider: FilesystemProvider = {}
      const args = { path: 'test.txt', operations: [{ search: 'old', replace: 'new' }] }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
    })

    it('should return error when file not found', async () => {
      const provider = createProvider(undefined)
      const args = { path: 'missing.txt', operations: [{ search: 'old', replace: 'new' }] }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
    })

    it('should return error when no operations provided', async () => {
      const provider = createProvider('content')
      const args = { path: 'test.txt', operations: [] }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
    })

    it('should handle single operation successfully', async () => {
      let writtenContent = ''
      const provider: FilesystemProvider = {
        readFile: async () => 'Hello world',
        writeFile: async (path, content) => {
          writtenContent = content
        },
      }
      const args = {
        path: 'test.txt',
        operations: {
          search: 'world',
          replace: 'beautiful world',
        },
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
      expect(writtenContent).toBe('Hello beautiful world')
    })

    it('should handle multiple operations successfully', async () => {
      let writtenContent = ''
      const provider: FilesystemProvider = {
        readFile: async () => 'function test() {\n  return 42;\n}',
        writeFile: async (path, content) => {
          writtenContent = content
        },
      }
      const args = {
        path: 'test.ts',
        operations: [
          {
            search: 'function test() {',
            replace: 'function test() {\n  console.log("debug");',
          },
          {
            search: 'return 42;',
            replace: 'return 42;\n  // Added comment',
          },
        ],
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
      expect(writtenContent).toBe('function test() {\n  console.log("debug");\n  return 42;\n  // Added comment\n}')
    })

    it('should handle start of file marker', async () => {
      let writtenContent = ''
      const provider: FilesystemProvider = {
        readFile: async () => 'export const value = 42;',
        writeFile: async (path, content) => {
          writtenContent = content
        },
      }
      const args = {
        path: 'test.ts',
        operations: {
          search: '<<<START_OF_FILE>>>',
          replace: '// Header comment\n',
        },
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
      expect(writtenContent).toBe('// Header comment\nexport const value = 42;')
    })

    it('should handle end of file marker', async () => {
      let writtenContent = ''
      const provider: FilesystemProvider = {
        readFile: async () => 'export const value = 42;',
        writeFile: async (path, content) => {
          writtenContent = content
        },
      }
      const args = {
        path: 'test.ts',
        operations: {
          search: '<<<END_OF_FILE>>>',
          replace: '\n// Footer comment',
        },
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
      expect(writtenContent).toBe('export const value = 42;\n// Footer comment')
    })

    it('should return error when text not found', async () => {
      const provider = createProvider('Hello world')
      const args = {
        path: 'test.txt',
        operations: {
          search: 'missing text',
          replace: 'replacement',
        },
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
    })

    it('should replace only first occurrence', async () => {
      let writtenContent = ''
      const provider: FilesystemProvider = {
        readFile: async () => 'test test test',
        writeFile: async (path, content) => {
          writtenContent = content
        },
      }
      const args = {
        path: 'test.txt',
        operations: {
          search: 'test',
          replace: 'replaced',
        },
      }

      const result = await handler(provider, args)

      expect(result).toMatchSnapshot()
      expect(writtenContent).toBe('replaced test test')
    })
  })
})
