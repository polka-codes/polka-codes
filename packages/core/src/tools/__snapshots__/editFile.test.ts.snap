// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`editFile tool toolInfo should have correct tool definition 1`] = `
{
  "description": "Request to edit file contents using search/replace operations. Supports multiple edit operations in a single call.",
  "examples": [
    {
      "description": "Replace specific text",
      "parameters": [
        {
          "name": "path",
          "value": "src/main.ts",
        },
        {
          "name": "operations",
          "value": {
            "replace": 
"function newFunction() {
  return "new implementation";
}"
,
            "search": 
"function oldFunction() {
  return 42;
}"
,
          },
        },
      ],
    },
    {
      "description": "Insert at start of file",
      "parameters": [
        {
          "name": "path",
          "value": "src/header.ts",
        },
        {
          "name": "operations",
          "value": {
            "replace": 
"// File header comment
"
,
            "search": "<<<START_OF_FILE>>>",
          },
        },
      ],
    },
    {
      "description": "Multiple operations in one call",
      "parameters": [
        {
          "name": "path",
          "value": "src/utils.ts",
        },
        {
          "name": "operations",
          "value": [
            {
              "replace": "import React, { useState } from "react"",
              "search": "import React from "react"",
            },
            {
              "replace": 
"function Component() {
  const [state, setState] = useState(false);
  return ("
,
              "search": 
"function Component() {
  return ("
,
            },
          ],
        },
      ],
    },
    {
      "description": "Append content at end of file",
      "parameters": [
        {
          "name": "path",
          "value": "src/footer.ts",
        },
        {
          "name": "operations",
          "value": {
            "replace": 
"
// End of file appended comment
"
,
            "search": "<<<END_OF_FILE>>>",
          },
        },
      ],
    },
  ],
  "name": "edit_file",
  "parameters": [
    {
      "description": "The path of the file to edit",
      "name": "path",
      "required": true,
      "usageValue": "File path here",
    },
    {
      "allowMultiple": true,
      "children": [
        {
          "description": "Text to search for and replace (use <<<START_OF_FILE>>> to insert at file start, <<<END_OF_FILE>>> to insert at file end)",
          "name": "search",
          "required": true,
          "usageValue": "Text to find and replace",
        },
        {
          "description": "Text to replace the search text with",
          "name": "replace",
          "required": true,
          "usageValue": "Replacement text",
        },
      ],
      "description": "Edit operation with search and replace parameters",
      "name": "operations",
      "required": true,
      "usageValue": "operations here",
    },
  ],
  "permissionLevel": 2,
}
`;

exports[`editFile tool handler should return error when provider lacks capabilities 1`] = `
{
  "message": "Not possible to edit file. Abort.",
  "type": "Error",
}
`;

exports[`editFile tool handler should return error when file not found 1`] = `
{
  "message": "<error><edit_file_path>missing.txt</edit_file_path><error_message>Could not find text: old</error_message></error>",
  "type": "Error",
}
`;

exports[`editFile tool handler should return error when no operations provided 1`] = `
{
  "message": "<error><edit_file_path>test.txt</edit_file_path><error_message>At least one edit operation is required</error_message></error>",
  "type": "Error",
}
`;

exports[`editFile tool handler should handle single operation successfully 1`] = `
{
  "message": "<edit_file_path>test.txt</edit_file_path>",
  "type": "Reply",
}
`;

exports[`editFile tool handler should handle multiple operations successfully 1`] = `
{
  "message": "<edit_file_path>test.ts</edit_file_path>",
  "type": "Reply",
}
`;

exports[`editFile tool handler should handle start of file marker 1`] = `
{
  "message": "<edit_file_path>test.ts</edit_file_path>",
  "type": "Reply",
}
`;

exports[`editFile tool handler should handle end of file marker 1`] = `
{
  "message": "<edit_file_path>test.ts</edit_file_path>",
  "type": "Reply",
}
`;

exports[`editFile tool handler should return error when text not found 1`] = `
{
  "message": "<error><edit_file_path>test.txt</edit_file_path><error_message>Could not find text: missing text</error_message></error>",
  "type": "Error",
}
`;

exports[`editFile tool handler should replace only first occurrence 1`] = `
{
  "message": "<edit_file_path>test.txt</edit_file_path>",
  "type": "Reply",
}
`;
